### PR TITLE
Access to email account in offline mode

### LINKED ISSUE #409: Access to email account in offline mode
As a customer I want access to the offline mode without being logged in in order to see and edit previously retrieved mails (that are already disc)

### LINKED ISSUE #410: Login possibility when online
As a customer I want to have the possibility to login my account when I am online again in order to access the online functionality without reopening the IMAP-Client 

    Acceptance criteria: 

    - WHEN I press the update button on the navigation panel 
    - AND my account is not logged in 
    - THEN a Login Dialog Window is opened and the login process is started 
    - WHEN Login successful 
    - THEN account is logged in and update process started

### PR DESCRIPTION
**Pull request by elenagensch**
## What does this change?

[Feature] Adds the possibility to access offline mode of IMAP Client without being logged in. 

Splits Login Process from ICFolderDialog initialization. Emails and Folders are loaded from File Storage (structure from last synchronization). Uses the UIManager for password request. Layered structure for calls from UI to interface with returns implemented. Bugs for login fixed (fatal error extracted from login process).
Closes #409 and closes #410

Co-authored-by: Elena Gensch <elena.gensch@student.hpi.de>
Co-authored-by: Jan Groeneveld <jan.niklas.groeneveld@gmail.com>

## Decisions / Choices I made

We removed the Password Dialog, as it oppposes the layered structure of handling requests and there's a lot of logic (passing arround FolderDialog instance, defining button) just for retrieving the password. Uses UIManager instead.

## Checklist

- [X] Your branch has no merge conflicts with master

### CODE REVIEW

File: packages/IMAPClient-UI.package/ICFolderDialog.class/instance/initialize.st
Comment (elenagensch) [inline packages/IMAPClient-UI.package/ICFolderDialog.class/instance/initialize.st:5]: these two lines affect only ICEndpoint, should be moved there

File: packages/IMAPClient-UI.package/ICFolderDialog.class/instance/switchToNewEndpoint.st
Comment (elenagensch) [inline packages/IMAPClient-UI.package/ICFolderDialog.class/instance/switchToNewEndpoint.st:6]: folderWith triggers server communication (only possible when logged in)

File: packages/IMAPClient-UI.package/ICFolderDialog.class/instance/switchToNewEndpoint.st
Comment (elenagensch) [inline packages/IMAPClient-UI.package/ICFolderDialog.class/instance/switchToNewEndpoint.st:10]: remove newline

**Review by laugengebaeck — COMMENTED** [review]
Looks good, but I don't really see where the login happens now (the `ICFolderDialog` opens without login and the update button also does not seem to trigger login). Can you explain this?

**Comment by elenagensch** [comment]
We are still working on that, this adresses only #409

**Review by laugengebaeck — COMMENTED** [review]
The new changes look good! 

But when trying the feature locally, I currently get a `MessageNotUnderstood: UndefinedObject>>do:` error. Is this expected behaviour? :sweat_smile:

**Comment by elenagensch** [comment]
No it's not, we still need to find a solution to embed login in the update process

File: packages/IMAPClient-UI.package/ICFolderDialog.class/instance/login..st
Comment (laugengebaeck) [inline packages/IMAPClient-UI.package/ICFolderDialog.class/instance/login..st]: For this, probably the original (commented out) `testAccountWith` should be restored (currently returns constant true).

File: packages/IMAPClient-UI.package/ICFolderDialog.class/instance/login..st
Comment (laugengebaeck) [inline packages/IMAPClient-UI.package/ICFolderDialog.class/instance/login..st]: If you call the variable `continueAsking`, I would have expected a `whileTrue` loop. Maybe rename that variable so that the code is more intuitive.

File: packages/IMAPClient-UI.package/ICFolderDialog.class/instance/update.st
Comment (laugengebaeck) [inline packages/IMAPClient-UI.package/ICFolderDialog.class/instance/update.st]: probably no point at the end (according to code standards)

File: packages/IMAPClient-Core.package/ICEndPoint.class/instance/testAccountWith..st
Comment (laugengebaeck) [inline packages/IMAPClient-Core.package/ICEndPoint.class/instance/testAccountWith..st]: should probably not constantly return true in the merged version

**Review by laugengebaeck — COMMENTED** [review]
Looks good (with some minor nitpicks from me)! I'm really happy that this feature finally works. Some commented out things and transcript outputs have to be removed before merging, but you probably know that :)

Edit: Also, `throwErrorIfAccountInvalid` in the `ICConnectionHandler` doesn't seem to be used anymore, so please remove it.

File: packages/IMAPClient-Core.package/ICEndPoint.class/instance/testAccountWith..st
Comment (blackjack2693) [reply packages/IMAPClient-Core.package/ICEndPoint.class/instance/testAccountWith..st]: that is dead code, will be removed (source: jan)

File: packages/IMAPClient-UI.package/ICFolderDialog.class/instance/login..st
Comment (blackjack2693) [reply packages/IMAPClient-UI.package/ICFolderDialog.class/instance/login..st]: thank you for the comment, I almost did not get it. It should be addressed.

File: packages/IMAPClient-UI.package/ICFolderDialog.class/instance/update.st
Comment (blackjack2693) [reply packages/IMAPClient-UI.package/ICFolderDialog.class/instance/update.st]: will be fixed

**Review by blackjack2693 — APPROVED** [review]
Looks great, will finally reduce the hard to read parts which were the login-process while adding new functions and saving huge amounts of code. Great!

File: packages/IMAPClient-UI.package/ICFolderDialog.class/instance/login..st
Comment (sm1lla) [inline packages/IMAPClient-UI.package/ICFolderDialog.class/instance/login..st]: last point not needed

**Review by sm1lla — APPROVED** [review]
As Lukas already pointed out, there are still some commented out parts that should be removed. Apart from that the code looks good!

**Review by laugengebaeck — APPROVED** [review]

**Comment by sm1lla** [comment]
Is it correct that there are several accounts named 'testAccountName' under the account I'm logged in with?

**Review by sm1lla — CHANGES_REQUESTED** [review]
I'm experiencing a few bugs. Please don't merge before we fixed this.

**Comment by laugengebaeck** [comment]
> Is it correct that there are several accounts named 'testAccountName' under the account I'm logged in with?

That's because I didn't tear down one of the new tests correctly (which is now fixed) :sweat_smile: So it shouldn't be related to this branch and you can just delete these accounts.

**Review by sm1lla — APPROVED** [review]

**Review by blackjack2693 — APPROVED** [review]

**Pull request merged**
